File: common/src/main/resources/ts/services/ResponseService.ts
import {idiom, ng, notify} from 'entcore';
import http from 'axios';
import {
    FormElement,
    IResponseResponse,
    Question,
    Response,
    Responses,
    Section
} from "../models";
import {DataUtils} from "../utils";
import {Exports} from "@common/core/enums";
export interface ResponseService {
    list(question: Question, nbLines: number) : Promise<any>;
    listByForm(formId: number) : Promise<any>;
    listMineByDistribution(questionId: number, distributionId: number) : Promise<any>;
    listMineByDistributionAndQuestions(questionIds: number[], distributionId: number) : Promise<Response[]>;
    listByDistribution(distributionId: number) : Promise<any>;
    countByFormElement(formElement: FormElement) : Promise<any>;
    save(response: Response, questionType: number) : Promise<any>;
    saveMultiple(mapQuestionResponsesToSave: Map<Question, Response[]>, distributionId: number) : Promise<Response[]>;
    create(response: Response) : Promise<any>;
    createMultiple(distributionId: number, responses: Response[]) : Promise<Response[]>;
    update(response: Response) : Promise<any>;
    updateMultiple(distributionId: number, responses: Response[]) : Promise<Response[]>;
    delete(formId: number, responses: Response[]) : Promise<any>;
    deleteByQuestionAndDistribution(questionId: number, distributionId: number) : Promise<any>;
    deleteMultipleByQuestionAndDistribution(questionIds: number[], distributionId: number) : Promise<any>;
    export(formId: number, type: string, images?: any) : Promise<any>;
}

export const responseService: ResponseService = {
    async list(question: Question, nbLines: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${question.id}/responses?nbLines=${nbLines}&formId=${question.form_id}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));

            throw err;
        }
    },

    async listByForm(formId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/forms/${formId}/responses`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async listMineByDistribution(questionId: number, distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${questionId}/distributions/${distributionId}/responses`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async listMineByDistributionAndQuestions(questionIds: number[], distributionId: number) : Promise<Response[]> {
        try {
            let data: IResponseResponse[] = DataUtils.getData(await http.get(`/formulaire/distributions/${distributionId}/responses/multiple`, { params: questionIds }));
            return data.map((rr: IResponseResponse) => new Response().build(rr));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async listByDistribution(distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/distributions/${distributionId}/responses`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.list'));
            throw err;
        }
    },

    async countByFormElement (formElement: FormElement) : Promise<any> {
        try {
            let questionIds = [];
            if (formElement instanceof Section) {
                for (let question of formElement.questions.all) {
                    questionIds.push(question.id);
                }
            }
            else if (formElement instanceof Question) {
                questionIds.push(formElement.id);
            }
            return DataUtils.getData(await http.get(`/formulaire/responses/count`, { params: questionIds }));
        } catch(err) {
            notify.error(idiom.translate('formulaire.error.responseService.get'));
            throw err;
        }
    },

    async save(response: Response, questionType: number) : Promise<any> {
        response.formatBeforeSaving(questionType);
        return response.id ? await this.update(response) : await this.create(response);
    },

    async saveMultiple(mapQuestionResponsesToSave: Map<Question, Response[]>, distributionId: number) : Promise<Response[]> {
        if (!mapQuestionResponsesToSave || mapQuestionResponsesToSave.size <= 0) {
            return [];
        }

        let  responsesToUpdate: Response[] = [];
        let responsesToCreate: Response[] = [];

        mapQuestionResponsesToSave.forEach((responses: Response[], question: Question) => {
            for (let response of responses) {
                response.formatBeforeSaving(question.question_type);
                response.id ? responsesToUpdate.push(response) : responsesToCreate.push(response);
            }
        });

        let promises: Promise<Response[]>[] = [this.updateMultiple(distributionId, responsesToUpdate), this.createMultiple(distributionId, responsesToCreate)];
        try {
            let results: Response[][] = await Promise.all(promises);
            return results[0].concat(results[1]);
        } catch (err) {
            notify.error(idiom.translate( 'formulaire.error.questionService.save'));
            throw err;
        }
    },

    async create(response: Response) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/questions/${response.question_id}/responses`, response));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.create'));
            throw err;
        }
    },

    async createMultiple(distributionId: number, responses: Response[]) : Promise<Response[]> {
        try {
            if (!responses || responses.length <= 0) {
                return [];
            }
            let data: IResponseResponse[] = DataUtils.getData(await http.post(`/formulaire/distributions/${distributionId}/responses/multiple`, responses));
            return data.map((rr: IResponseResponse) => new Response().build(rr));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.create'));
            throw err;
        }
    },

    async update(response: Response) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/responses/${response.id}`, response));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.update'));
            throw err;
        }
    },

    async updateMultiple(distributionId: number, responses: Response[]) : Promise<Response[]> {
        try {
            if (!responses || responses.length <= 0) {
                return [];
            }
            let data: IResponseResponse[] = DataUtils.getData(await http.put(`/formulaire/distributions/${distributionId}/responses`, responses));
            return data.map((rr: IResponseResponse) => new Response().build(rr));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.update'));
            throw err;
        }
    },

    async delete(formId, responses) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/responses/${formId}`, { data: responses } ));
        } catch (e) {
            notify.error(idiom.translate('formulaire.error.responseService.delete'));
            throw e;
        }
    },

    async deleteByQuestionAndDistribution(questionId: number, distributionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/responses/${distributionId}/questions/${questionId}`));
        } catch (e) {
            notify.error(idiom.translate('formulaire.error.responseService.delete'));
            throw e;
        }
    },

    async deleteMultipleByQuestionAndDistribution(questionIds: number[], distributionId: number) : Promise<any> {
        try {
            if (!questionIds || questionIds.length <= 0) {
                return [];
            }
            return DataUtils.getData(await http.delete(`/formulaire/responses/${distributionId}/questions`, { data: questionIds }));
        } catch (e) {
            notify.error(idiom.translate('formulaire.error.responseService.delete'));
            throw e;
        }
    },

    async export(formId: number, type: string, images: any) : Promise<any> {
        try {
            if (type === Exports.CSV) {
                return await http.post(`/formulaire/responses/export/${formId}/csv`, {});
            }
            else if (type === Exports.PDF) {
                return await http.post(`/formulaire/responses/export/${formId}/pdf`, images, { responseType: "arraybuffer" });
            }
            else {
                notify.error(idiom.translate('formulaire.error.responseService.export'));
            }
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseService.export'));
            throw err;
        }
    }
};

export const ResponseService = ng.service('ResponseService', (): ResponseService => responseService);

File: formulaire/src/main/java/fr/openent/formulaire/service/ResponseService.java
package fr.openent.formulaire.service;

import fr.openent.form.core.models.Response;
import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

import java.util.List;

public interface ResponseService {
    /**
     * List all the responses to a specific question
     * @param questionId question identifier
     * @param nbLines number of lines already loaded
     * @param handler function handler returning JsonArray data
     */
    void list(String questionId, String nbLines, JsonArray distribs, Handler<Either<String, JsonArray>> handler);

    /**
     * List all my responses to a specific question for a specific distribution
     * @param questionId question identifier
     * @param distributionId distribution identifier
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void listMineByDistribution(String questionId, String distributionId, UserInfos user, Handler<Either<String, JsonArray>> handler);


    /**
     * @param questionsIds array of questions identifiers
     * @param distributionId distribution identifier
     * @param userId  connected user's id
     */
    Future<List<Response>> listMineByQuestionsIds(JsonArray questionsIds, String distributionId, String userId);

    /**
     * List all responses for a specific distribution
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void listByDistribution(String distributionId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all responses for a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void listByForm(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all responses for specific ids
     * @param responseIds response identifiers
     */
    Future<List<Response>> listByIds(List<String> responseIds);

    /**
     * Count all the responses to a list of questions
     * @param questionIds questions identifiers
     * @param handler function handler returning JsonObject data
     */
    void countByQuestions(JsonArray questionIds, Handler<Either<String, JsonObject>> handler);

    /**
     * Get a specific response by id
     * @param responseId response identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String responseId, Handler<Either<String, JsonObject>> handler);

    /**
     * Get question ids of a specific form where specific distribution has missing responses
     * @param formId form identifier
     * @param distributionId distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void getMissingResponses(String formId, String distributionId, Handler<Either<String, JsonArray>> handler);

    /**
     * Create a response
     * @param response JsonObject data
     * @param user user connected
     * @param questionId question identifier
     * @param handler function handler returning JsonObject data
     */
    void create(JsonObject response, UserInfos user, String questionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create multiple responses
     * @param responses responses to create
     * @param userId    id of the connected user
     */
    Future<List<Response>> createMultiple(List<Response> responses, String userId);

    /**
     * Update a specific response
     * @param user user connected
     * @param responseId response identifier
     * @param response JsonObject data
     * @param handler function handler returning JsonObject data
     */
    void update(UserInfos user, String responseId, JsonObject response, Handler<Either<String, JsonObject>> handler);

    /**
     * Update multiple responses
     * @param responses responses to update
     * @param userId    id of the connected user
     */
    Future<List<Response>> updateMultiple(List<Response> responses, String userId);

    /**
     * Delete specific responses
     * @param responseIds responses identifiers
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void delete(JsonArray responseIds, String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete responses of a specific question of a specific distribution
     * @param questionId question identifier
     * @param distributionId distribution identifier
     * @param user user connected
     * @param handler function handler returning JsonArray data
     */
    void deleteByQuestionAndDistribution(String questionId, String distributionId, UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete responses for multiple questions of a specific distribution
     * @param questionIds list of question identifiers
     * @param distributionId distribution identifier
     * @return Future<Void>
     */
    Future<Void> deleteByQuestionsAndDistribution(List<Long> questionIds, String distributionId);

    /**
     * Delete responses by distribution id
     * @param String distribution identifier
     * @param handler function handler returning JsonArray data
     */
    void deleteAllByDistribution(String String, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responders of a specific form to export into a CSV file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void getExportCSVResponders(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responses of a specific form to export into a CSV file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void exportCSVResponses(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get all the responses of a specific form to export into a PDF file
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void exportPDFResponses(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete responses too old according to lifetime RGPD chosen
     * @param distributionIds id of deleted distributions
     * @param handler function handler returning JsonObject data
     */
    void deleteOldResponse(JsonArray distributionIds, Handler<Either<String, JsonArray>> handler);


}